README.md
# 🏴‍☠️🤖 Threat Intelligence Teams Bot

TITB is a fork from [Threat Intelligence Discord Bot from vx-underground](https://github.com/vxunderground/ThreatIntelligenceDiscordBot/) but for Microsoft Teams and modified to work as an hourly Github-Action

> The vx-underground Threat Intelligence Discord Bot gets updates from various clearnet domains, ransomware threat actor domains This bot will check for updates in intervals of 1800 seconds.

[![MIT License](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE) ![Version](https://img.shields.io/badge/version-2.5.0-blue.svg)  [![Twitter: JMousqueton](https://img.shields.io/twitter/follow/JMousqueton.svg?style=social)](https://twitter.com/JMousqueton) [![Last Run](https://github.com/JMousqueton/CTI-MSTeams-Bot/actions/workflows/fetchCTI.yml/badge.svg)](.github/workflows/fetchCTI.yml)  [![CodeQL](https://github.com/JMousqueton/CTI-MSTeams-Bot/actions/workflows/codeql-analysis.yml/badge.svg)](.github/workflows/codeql-analysis.yml)

## Description

* Written in Python

   ⚠️ required version Python 3.10+
* Requires Teams Webhook

Threat Intelligence Teams Bot gets updates from various clearnet domains and ransomware threat actor domains.

This bot will check for updates every 30 minutes.

The change I've made :

* GitHub-Action : see the [fetchCTI.yml](.github/workflows/fetchCTI.yml) file.
* Feeds List is externalized from the source code to the file [Feed.csv](Feed.csv)
* Use JSON lib to get ransomware attacks list from [Ransomwatch](https://ransomwatch.mousqueton.io)
* Add an emoji in front of the MS Teams Card related to some sources
* Use only one MS Teams Channel
* Refactoring with best practices
* No need to add any entry in [Config.txt](Config.txt) (automaticaly add)
* Check that python 3.10+ is present (needed for some functions)
* Add a [requirements.txt](requirements.txt)
* Add a [feedCheck.py](checkFeed.py) script to check the health of the feed from [Feed.csv](Feed.csv) file
* Add Options for command line [usage](#usage)
* Check if a new version is available
* Add [new sources](#sources)

I've decided to remove the TelegramBot because it was not relevant for my needs.

![](Screenshot.png)

## Installation

Clone the repository or download the [latest release](https://github.com/JMousqueton/CTI-MSTeams-Bot/releases/latest)

```bash
git clone https://github.com/JMousqueton/CTI-MSTeams-Bot
```

Install all the modules in ```requirements.txt```

```bash
pip3 install -r requirements.txt
```

## Configuration

### Github Action

* Create a MS-Teams WebHook  
* in an environment you will called `CI`, paste the created webhook url in a `MSTEAMS_WEBHOOK` variable.

### On a server (Windows, MacOS, Linux)

* Create a variable called ```MSTEAMS_WEBHOOK``` with the webhook URL

Example:

```bash
MSTEAMS_WEBHOOK=https://mousqueton.webhook.office.com/webhookb2/08589F1C-EEA2-4C92-A08B-66E59692FDE3/IncomingWebhook/3DEFFDD9-F3A8-4351-BDA7-142FAFB7473A
python3 TeamIntelBot.py -r -d 
```

* Schedule the script for example every hours via the crontab

*Note: the IDs have been generated with uuidgen for example purpose* 😛

## Usage

```bash
python3 TeamsIntelBot.py -h
Usage: TeamsIntelBot.py [options]

Options:
  --version       show program's version number and exit
  -h, --help      show this help message and exit
  -q, --quiet     Quiet mode
  -D, --debug     Debug mode : only output on screen nothing send to MS Teams
  -d, --domain    Enable Red Flag Domains source
  -r, --reminder  Enable monthly reminder of Feeds
```

- For french user, I recommand using flags -d and -r

```python3 TeamIntelBot.py -r -d```

- For other, only flag -r

```python3 TeamIntelBot.py -r```

### Proxy

If you use a proxy don't forget to use the proxies variables:

```bash
set https_proxy=http://x.x.x.x:port
set http_proxy=http://x.x.x.x:port
```

I've also add a script called ```checkFeed.py``` to check if feeds are valid and what is the last published date. This script read the ```Feed.csv``` file.

```bash
python3 checkFeed.py 

✅ Modexp (Sun, 31 Jul 2022 00:01:53 +0000)
✅ James Forshaw (2022-07-16T21:49:00.000-07:00)
✅ Adam Chester (Sat, 09 Jul 2022 23:00:00 GMT)
✅ Microsoft Security (Thu, 11 Aug 2022 16:00:00 +0000)
✅ Recorded Future (Thu, 18 Aug 2022 00:00:00 GMT)
✅ SentinelOne (Wed, 11 May 2022 14:56:53 +0000)
✅ RedCanary (Thu, 18 Aug 2022 21:53:55 +0000)
✅ Cyber-News (Fri, 19 Aug 2022 15:14:56 +0000)
✅ Leak-Lookup (Fri, 19 Aug 2022 04:00:02 +0200)
✅ ATT (2022-08-17T10:00:00+00:00)
✅ US-CERT CISA (Tue, 16 Aug 2022 15:38:42 +0000)
✅ NCSC (Thu, 18 Aug 2022 23:00:00 GMT)
✅ Center of Internet Security (Thu, 18 Aug 2022 01:43:07 -0400)
✅ FR-CERT Alertes (Tue, 31 May 2022 11:12:01 +0000)
✅ FR-CERT Avis (Fri, 19 Aug 2022 11:22:29 +0000)
✅ EU-ENISA Publications (2022-07-27T10:00:00Z)
✅ Microsoft Sentinel (Thu, 18 Aug 2022 08:31:51 PDT)
```

## Adding or removing RSS Feeds to monitor
All monitored RSS feeds are in [Feed.csv](Feed.csv) file. To add a new RSS feed simply append a new entry. e.g.

In the ```Feed.csv``` file:

```text
https://grahamcluley.com/feed/,Graham Cluley
https://1337WebsiteIWannaFollow.com/feed/,1337Website
```

## Sources

I've added the following sources :

* 🇫🇷 FR-CERT Avis (aka [ANSSI](https://www.ssi.gouv.fr/)) : notifications from gov French CERT
* 🇫🇷 FR-CERT Alertes (aka [ANSSI](https://www.ssi.gouv.fr/)) : Alerts from gov French CERT
* [Leak-lookup](https://leak-lookup.com/) : Leak notification
* [Cyber-News](https://www.cyber-news.fr)
* ATT CyberSecurity Blog
* 🇪🇺 ENSIA Publications
* NCC Group
* Microsoft Sentinel
* SANS
* [Red Flag Domains](https://red.flag.domains/) ⚠️ You shoudl use -d flag to enable this source dedicated to France
* [Google TAG](https://blog.google/threat-analysis-group/)  

## ToDo

* ~~Create a flag to activate or not the Red Flag Domains source because it's only for french~~ (released in version 2.3)
* ~~Disable line with # in [Feed.csv](Feed.csv)~~ (in version 2.4 not released yet)

## Credit

This was made by smelly__vx over a slow and boring weekend. We hope it provides some value to your channel and/or organization.

Thanks to my fellow students from [🏴‍☠️ Ecole 2600](https://www.ecole2600.com) for the support and advice during nights 😛

Thanks to the current users of this Bot who help me to improve it

Thanks to Olivier for the proxy documentation 🍻


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.5.0]- 2023-01-11

### Changed

- Rewrite the red flag parser

## [2.4.0]- 2022-10-16

### Changed

- Replace Google feed by Google TAG feed
- Make Emoji a function  
- Unit42 emoji

### Add

- Add Google TAG emoji
- Add MSRC Security Update feed
- Add MSRC Security Update emoji
- Don't check Feed lines with # in [CheckFeed.py](checkFeed.py) and [TeamsIntelBot.py](TeamsIntelBot.py)
- Monthly Feed Reminder (with corrected bug)

## [2.3.0]- 2022-10-15

### Changed

- Some minor changes

### Add

- None

## [2.2.0 UNRELEASED]- 2022-08-29

### Changed

- Code optimizations

### Add

- French flag for ransomware domain hit in .fr
- Unit42 Feed
- DataBreaches Feed (disabled by default)

## [2.1.2 UNRELEASED]- 2022-08-23

### Changed

- Modify Config.txt format
- Modify the order of reading RSS Feed

### Add

- None

## [2.1.0 UNRELEASED]- 2022-08-22

### Changed

- Add a flag to enable Red Flag Domains source

### Add

- None

## [2.0.2] - 2022-08-21

### Changed

- Correct a bug in the debug condition

### Add

- None

## [2.0.0] - 2022-08-21

### Changed

- None

### Add

- Options for CLI
  - Debug mode for screen output only
  - Quiet mode
- Check python version in the script
- And much more :)


checkFeed.py
import urllib.request
import feedparser,csv

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'

with open('Feed.csv', newline='') as f:
    reader = csv.reader(f)
    RssFeedList = list(reader)

for RssItem in RssFeedList:
    if '#' in str(RssItem[0]):
                continue
    Feed = feedparser.parse(RssItem[0])
    try:
        print("✅ " + color.BOLD + RssItem[1] + color.END + " ("+ Feed.entries[0].published + ")")
    except:
        try:
            print("✅ " + color.BOLD + RssItem[1] + color.END + " ("+ Feed.entries[0].updated + ")")
        except:
            print("❌ " + color.BOLD  + RssItem[1] +  color.END + color.RED + " RSS ERROR !" + color.END)

TeamsIntelBot.py
#!/usr/bin/env python3  
# -*- coding: utf-8 -*- 
#----------------------------------------------------------------------------
# Created By  : Julien Mousqueton @JMousqueton
# Original By : VX-Underground 
# Created Date: 22/08/2022
# Version     : 2.5.0 (2023-01-11)
# ---------------------------------------------------------------------------

# ---------------------------------------------------------------------------
# Imports 
# ---------------------------------------------------------------------------
import feedparser
import time, requests
import csv # Feed.csv
import sys # Python version 
import json # Ransomware feed via ransomwatch 
from configparser import ConfigParser
import os # Webhook OS Variable and Github action 
from os.path import exists
from optparse import OptionParser
import urllib.request
from bs4 import BeautifulSoup
from datetime import datetime, timedelta
import re

# ---------------------------------------------------------------------------
# Function to send MS-Teams card 
# ---------------------------------------------------------------------------
def Send_Teams(webhook_url:str, content:str, title:str, color:str="000000") -> int:
    """
      - Send a teams notification to the desired webhook_url
      - Returns the status code of the HTTP request
        - webhook_url : the url you got from the teams webhook configuration
        - content : your formatted notification content
        - title : the message that'll be displayed as title, and on phone notifications
        - color (optional) : hexadecimal code of the notification's top line color, default corresponds to black
    """
    response = requests.post(
        url=webhook_url,
        headers={"Content-Type": "application/json"},
        json={
            "themeColor": color,
            "summary": title,
            "sections": [{
                "activityTitle": title,
                "activitySubtitle": content
            }],
        },
    )
    return response.status_code # Should be 200

# ---------------------------------------------------------------------------
# Fetch Ransomware attacks from https://ransomwatch.mousqueton.io  
# ---------------------------------------------------------------------------
def GetRansomwareUpdates():
    
    Data = requests.get("https://raw.githubusercontent.com/jmousqueton/ransomwatch/main/posts.json")
        
    for Entries in Data.json():

        DateActivity = Entries["discovered"]
            
        # Correction for issue #1 : https://github.com/JMousqueton/CTI-MSTeams-Bot/issues/1
        try:
            TmpObject = FileConfig.get('Ransomware', Entries["group_name"])
        except:
            FileConfig.set('Ransomware', Entries["group_name"], " = ?")
            TmpObject = FileConfig.get('Ransomware', Entries["group_name"])

        if TmpObject.endswith("?"):
            FileConfig.set('Ransomware', Entries["group_name"], DateActivity)
        else:
            if(TmpObject >= DateActivity):
                continue
        #else:
        #    FileConfig.set('Ransomware', Entries["group_name"], Entries["discovered"])

        OutputMessage = "Group : <b>"
        OutputMessage += "<a href=\"https://www.ransomware.live/#/profiles?id="
        OutputMessage += Entries["group_name"]
        OutputMessage += "\">"
        OutputMessage += Entries["group_name"]
        OutputMessage += "</a>"
        OutputMessage += "</b><br>🗓 "
        OutputMessage += Entries["discovered"]
        OutputMessage += "</b><br>🌍 <a href=\"https://www.google.com/search?q="
        OutputMessage += Entries["post_title"].replace("*.", "")
        OutputMessage += "\">"
        OutputMessage += Entries["post_title"]
        OutputMessage += "</a>"
        
        Title = "🏴‍☠️ 🔒 "     

        if Entries["post_title"].find(".fr") != -1:
            Title += " 🇫🇷 "

        Title += Entries["post_title"].replace("*.", "") 
        Title += " by "
        Title += Entries["group_name"]

        if options.Debug:
            print(Entries["group_name"] + " = " + Title + " ("  + Entries["discovered"]+")")
        else:
            Send_Teams(Url,OutputMessage,Title)
            time.sleep(3)

        FileConfig.set('Ransomware', Entries["group_name"], Entries["discovered"])

    with open(ConfigurationFilePath, 'w') as FileHandle:
        FileConfig.write(FileHandle)


# ---------------------------------------------------------------------------
# Add nice Emoji in front of title   
# ---------------------------------------------------------------------------
def Emoji(feed):
    # Nice emoji :) 
    match feed:
        case "Leak-Lookup":
            Title = '💧 '
        case "VERSION":
            Title = '🔥 '
        case "DataBreaches":
            Title = '🕳 '
        case "FR-CERT Alertes" | "FR-CERT Avis":
            Title = '🇫🇷 '
        case "EU-ENISA Publications":
            Title = '🇪🇺 '
        case "Cyber-News":
            Title = '🕵🏻‍♂️ '
        case "Bleeping Computer":
            Title = '💻 '
        case "Microsoft Sentinel":
            Title = '🔭 '
        case "Hacker News":
            Title = '📰 '
        case "Cisco":
            Title = '📡 '
        case "Securelist":
            Title = '📜 '
        case "ATT":
            Title = '📞 '
        case "Google TAG":
            Title = '🔬 '
        case "DaVinci Forensics":
            Title = '📐 '
        case "VirusBulletin":
            Title = '🦠 '
        case "Information Security Magazine":
            Title = '🗞 '
        case "US-CERT CISA":
            Title = '🇺🇸 '
        case "NCSC":
            Title = '🇬🇧 '
        case "SANS":
            Title = '🌍 '
        case "malpedia":
            Title = '📖 '
        case "Unit42":
            Title = '🚓 '
        case "Microsoft Security":
            Title = 'Ⓜ️ '
        case "Checkpoint Research":
            Title = '🏁 '
        case "Proof Point":
            Title = '🧾 '
        case "RedCanary":
            Title = '🦆 '
        case "MSRC Security Update":
            Title = '🚨 '
        case "CIRCL Luxembourg":
            Title = '🇱🇺 '
        case _:
            Title = '📢 '
    return Title


# ---------------------------------------------------------------------------
# Function fetch RSS feeds  
# ---------------------------------------------------------------------------
def GetRssFromUrl(RssItem):
    NewsFeed = feedparser.parse(RssItem[0])
    DateActivity = ""
    IsInitialRun = False

    for RssObject in reversed(NewsFeed.entries):

        try:
            DateActivity = time.strftime('%Y-%m-%dT%H:%M:%S', RssObject.published_parsed)
        except: 
            DateActivity = time.strftime('%Y-%m-%dT%H:%M:%S', RssObject.updated_parsed)
        
        # Correction for issue #1 : https://github.com/JMousqueton/CTI-MSTeams-Bot/issues/1
        try:
            TmpObject = FileConfig.get('Rss', RssItem[1])
        except:
            FileConfig.set('Rss', RssItem[1], " = ?")
            TmpObject = FileConfig.get('Rss', RssItem[1])

        if TmpObject.endswith("?"):
            FileConfig.set('Rss', RssItem[1], DateActivity)
        else:
            if(TmpObject >= DateActivity):
                continue

        OutputMessage = "Date: " + DateActivity
        OutputMessage += "<br>"
        # OutputMessage += "Title:<b> " + RssObject.title
        OutputMessage += "Source:<b> " + RssItem[1]
        OutputMessage += "</b><br>"
        OutputMessage += "Read more: " + RssObject.link
        OutputMessage += "<br>"

        Title = Emoji(RssItem[1])
        Title += " " + RssObject.title

        if RssItem[1] == "VERSION":
                Title ='🔥 A NEW VERSION IS AVAILABLE : ' + RssObject.title
       
        if options.Debug:
            print(Title + " : " + RssObject.title + " (" + DateActivity + ")")
        else:
            Send_Teams(Url,OutputMessage,Title)
            time.sleep(3)
        
        FileConfig.set('Rss', RssItem[1], DateActivity)

    with open(ConfigurationFilePath, 'w') as FileHandle:
        FileConfig.write(FileHandle)

# ---------------------------------------------------------------------------
# Function fetch Red Flag domains 
# ---------------------------------------------------------------------------
def GetRedFlagDomains():
    now = datetime.now()
    format = "%Y-%m-%d"
    today = now.strftime(format)
    yesterday = now - timedelta(days=1)
    yesterday = yesterday.strftime(format)

    try:
        TmpObject = FileConfig.get('Misc',"redflagdomains")
    except:
        FileConfig.set('Misc', "redflagdomains", str(yesterday))
        TmpObject = str(yesterday)

    TmpObject = datetime.strptime(TmpObject, '%Y-%m-%d')
    today = datetime.strptime(today, '%Y-%m-%d')

    today = today.date()
    TmpObject = TmpObject.date()

    if(TmpObject < today):
        url="https://red.flag.domains/posts/"+ str(today) + "/"
        try:
            response = urllib.request.urlopen(url)
            soup = BeautifulSoup(response, 
                                'html.parser', 
                                from_encoding=response.info().get_param('charset'))
            # response_status = response.status
            #if soup.findAll("meta", property="og:description"):
            #    OutputMessage = soup.find("meta", property="og:description")["content"][4:].replace('.wf ','').replace('.yt ','').replace('.re ','').replace('[','').replace(']','')
            div = soup.find("div", {"class": "content", "itemprop": "articleBody"})
            for p in div.find_all("p"):
                OutputMessage = re.sub("[\[\]]", "", (p.get_text()))
            Title = "🚩 Red Flag Domains créés ce jour (" +  str(today) + ")"
            FileConfig.set('Misc', "redflagdomains", str(today))
            if options.Debug:
                print(Title)
                print(OutputMessage)
            else:
                Send_Teams(Url,OutputMessage.replace('\n','<br>'),Title)
                time.sleep(3)
        except:
            pass 
    with open(ConfigurationFilePath, 'w') as FileHandle:
        FileConfig.write(FileHandle)

# ---------------------------------------------------------------------------
# Function Send Feeds Reminder 
# ---------------------------------------------------------------------------
def SendReminder():
    now = datetime.now()
    format = "%Y-%m-%d"
    today = now.strftime(format)
    lastmonth = now - timedelta(days=31)
    lastmonth = lastmonth.strftime(format)
    try:
        TmpObject = FileConfig.get('Misc',"reminder")
    except:
        FileConfig.set('Misc', "reminder", str(lastmonth))
        TmpObject = str(lastmonth)
   
    TmpObject = datetime.strptime(TmpObject, '%Y-%m-%d')
    today = datetime.strptime(today, '%Y-%m-%d')
    lastmonth = datetime.strptime(lastmonth, '%Y-%m-%d')
    
    if(TmpObject < lastmonth):
        Title = "🤔 Monthly Feeds Reminder"
        if options.Debug:
            print(Title)
        OutputMessage="Feeds : "
        OutputMessage += "<br>"
        with open('Feed.csv', newline='') as f:
            reader = csv.reader(f)
            RssFeedList = list(reader)

        for RssItem in RssFeedList:
            if '#' in str(RssItem[0]):
                continue
            Feed = feedparser.parse(RssItem[0])
            try:
                OutputMessage += Emoji(RssItem[1]) + RssItem[1] + "  (" + Feed.entries[0].published + ")"
                OutputMessage += "<br>"
            except:
                try:
                    OutputMessage += Emoji(RssItem[1]) + RssItem[1] + "  (" + Feed.entries[0].updated + ")"
                    OutputMessage += "<br>"
                except:
                    continue
        if options.Domains: 
            OutputMessage += "Misc : "
            OutputMessage += "<br>"
            OutputMessage += "🚩 Red Flag Domains"
            OutputMessage += "<br>"
        OutputMessage += "Ransomware :"
        OutputMessage += "<br>"
        OutputMessage += "🏴‍☠️ 🔒 Ransomware Leaks"
        OutputMessage += "<br><br>"
        OutputMessage += "Coded with ❤️ by JMousqueton"
        OutputMessage += "<BR>"
        OutputMessage += "Code : https://github.com/JMousqueton/CTI-MSTeams-Bot"
        today = today.strftime(format)
        FileConfig.set('Misc', "reminder", str(today))
        if options.Debug:
            print(OutputMessage)
        else: 
            Send_Teams(Url,OutputMessage,Title)    

    with open(ConfigurationFilePath, 'w') as FileHandle:
        FileConfig.write(FileHandle)


# ---------------------------------------------------------------------------
# Log  
# ---------------------------------------------------------------------------
def CreateLogString(RssItem):
    LogString = "[*]" + time.ctime()
    LogString += " " + "checked " + RssItem
    if not options.Quiet: 
        print(LogString)
    time.sleep(2) 

# ---------------------------------------------------------------------------
# Main   
# ---------------------------------------------------------------------------    
if __name__ == '__main__':
    parser = OptionParser(usage="usage: %prog [options]",
                          version="%prog 2.2.0")
    parser.add_option("-q", "--quiet",
                      action="store_true",
                      dest="Quiet",
                      default=False,
                      help="Quiet mode")
    parser.add_option("-D", "--debug",
                      action="store_true", 
                      dest="Debug",
                      default=False,
                      help="Debug mode : only output on screen nothing send to MS Teams",)
    parser.add_option("-d", "--domain",
                      action="store_true", 
                      dest="Domains",
                      default=False,
                      help="Enable Red Flag Domains source",)
    parser.add_option("-r", "--reminder",
                      action="store_true",
                      dest="Reminder",
                      default=False,
                      help="Enable monthly reminder of Feeds")
    (options, args) = parser.parse_args()

    # Get Microsoft Teams Webhook from Github Action CI:Env.  
    Url=os.getenv('MSTEAMS_WEBHOOK')

    # expects the configuration file in the same directory as this script by default, replace if desired otherwise
    ConfigurationFilePath = os.path.join(os.path.split(os.path.abspath(__file__))[0], 'Config.txt')

    # Make some simple checks before starting 
    if sys.version_info < (3, 10):
        sys.exit("Please use Python 3.10+")
    if (str(Url) == "None" and not options.Debug):
             sys.exit("Please use a MSTEAMS_WEBHOOK variable")
    if not exists(ConfigurationFilePath):
        sys.exit("Please add a Config.txt file")
    if not exists("./Feed.csv"):
        sys.exit("Please add the Feed.cvs file")
    
    # Read the Config.txt file   
    # ConfigurationFilePath = "./Config.txt" ##path to configuration file
    FileConfig = ConfigParser()
    FileConfig.read(ConfigurationFilePath)

    with open('Feed.csv', newline='') as f:
        reader = csv.reader(f)
        RssFeedList = list(reader)
            
    for RssItem in RssFeedList:
        if '#' in str(RssItem[0]):
            continue
        GetRssFromUrl(RssItem)
        CreateLogString(RssItem[1])

    GetRansomwareUpdates()
    CreateLogString("Ransomware List")
    
    if options.Domains: 
        GetRedFlagDomains()
        CreateLogString("Red Flag Domains")

    if options.Reminder:
        SendReminder()
        CreateLogString("Reminder")
